v-next/hardhat-viem/src/accounts.ts

```diff
@@ -0,0 +1,22 @@
+import type { EthereumProvider } from "@ignored/hardhat-vnext/types/providers";
+import type * as viemT from "viem";
+
+const accountsCache = new Map<EthereumProvider, viemT.Address[]>();
```

Comment:
I am tempted here to say that we should associate the accounts cache as a private property of the `connection.viem` object, and pass it in as a parameter here - to avoid the global.

Or as an alternative use [WeakMap](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap), so that providers+cached accounts get garbage collected when the connection is no longer used.

Reply:
I don’t think we can pass it if it’s `private`, right? And I’m not a fan of exposing it as a `public` property. I like the idea of using a `WeakMap` though, since the concern is valid. We should apply it here and in the `chains.ts` file.

Reply:
I was thinking something along the lines of:

```ts
const accountsCache = ...

connection.viem = {
  _accountsCache: accountsCache
  getAccounts: () => {
    return getAccounts(provider, accountsCache)
  }
}
```

But I am happy with the WeakMap approach. Lets do that.

Reply:
Done [here](https://github.com/NomicFoundation/hardhat/pull/5825/commits/dde02e9c4d406f5413dcd39f7f9d393bd5307a8a).